NEW: Added InputDeviceMatcher.WithManufacturerContains(string noRegexMatch) to improve DualShockSupport initialization performance

AlexTyrer opened the pull request:
### Description

The **InputDeviceMatcher.WithManufacturer**() can perform a regex match even for simple patterns (eg "Sony.+Entertainment" and "PLAYSTATION(R)3 Controller" are both treated as Regexes) which is surprisingly costly on initialization.

This PR addresses that by adding an API to perform a simple string keyword match on the device manufacturer string,

### Changes made

The new **WithManufacturerContains**(string **noRegexMatch**) API performs a String.**Contains**() check instead looking for the simple pattern in the manufacturer string.

Changed the DualShockSupport for **DualShock4Gamepad** and **DualShock3Gamepad** to avoid regex matching.

These changes save 25ms in **DualShockSupport.Initialize** during player start-up (26ms -> 1ms)


### Testing

Local testing using SONY controllers and Quest VR devices.

### Risk

Previously a controller that reported the manufacturer as "SONY Entertainment" would match - with this change a hypothetical device that reported the manufacturer as "Entertainment Not SONY" would also match. I think this is a risk worth taking.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - JIRA ticket linked, example ([case %<ID>%](https://issuetracker.unity3d.com/product/unity/issues/guid/<ID>)). If it is a private issue, just add the case ID without a link.
    - Jira port for the next release set as "Resolved".
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [x] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [x] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

After merge:

- [ ] Create forward/backward port if needed. If you are blocked from creating a forward port now please add a task to ISX-1444.

AlexTyrer committed:
[InputSystem] Added WithManufacturerContains.WithManufacturerContains(string) - improve DualShockSupport initialization performance

o WithManufacturerContains() provides for simple keyword (no regex)  matching in the manufacturer string
o Changes DualShock4Gamepad / DualShock3Gamepad matchers to avoid regex creation on initialization

These changes save 25ms in DualShockSupport.Initialize during player start-up (26ms -> 1ms)

AlexTyrer committed:
[InputSystem] Improve parameter name: WithManufacturerContains(string noRegexMatch), added chanelog entry (ISX-1411)

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Devices/InputDeviceMatcher.cs:
```diff
@@ -142,6 +142,17 @@ public InputDeviceMatcher WithManufacturer(string pattern, bool supportRegex = t
+        /// <seealso cref="InputDeviceDescription.manufacturer"/>
+        public InputDeviceMatcher WithManufacturerContains(string noRegExPattern)
+        {
+            return With(kManufacturerContainsKey, noRegExPattern, supportRegex:false);
```
Given that supportRegex:false is passed, do you reallly need kManufacturerContainsKey.
It seems you could just use kManufacturerKey since the object type will remain a string

(but I may have missed something)

AlexTyrer replied:
The different key is needed to take the different path through MatchPercentage().

kManufacturerKey -> MatchSingleProperty -> string.Compare(...) == 0  // string/regex needs exactly match
kManufacturerContainsKey -> MatchSinglePropertyContains -> string.Contains(...) // string just needs to be in there somewhere

lyndon-unity commented on Packages/com.unity.inputsystem/InputSystem/Plugins/DualShock/DualShockSupport.cs:
```diff
@@ -65,8 +66,9 @@ public static void Initialize()
                     .WithInterface("HID")
-                    .WithManufacturer("Sony.+Entertainment")
-                    .WithProduct("PLAYSTATION(R)3 Controller"));
+                    .WithManufacturerContains("Sony")
```
I'm wondering it this could be just WithManufacturer("Sony", supportRegex:false) ?

There is a small chance this matches wider now - E.g. Entertainment Sony would match now, but I don't think this is likely to be any false positives.

AlexTyrer replied:
I believe it needs to be an exact match as the code does:

`string.Compare(...) == 0 `

... so we would need to provide the entire manufacturer string - and there seem to be two flavours of Sony manufacturer string:

- "Sony Interactive Entertainment"
- "Sony Computer Entertainment"

... hence the regex in the original code.

Just checked this.

`                    .WithManufacturer("Sony", supportRegex:false)   // test
`
... and some of the tests fail.

This is the reason why I added the 'Contains' thing - an explicit non-regex pattern that just has to be in the manufacturer string.

lyndon-unity reviewed (approved):
Generally I approve the change and the savings. 
I've asked a couple of questions though as it looks like a slightly smaller change could give the same result.

Might also be worth adding Lewis Hammond to this PR from the PlayStation team

AlexTyrer committed:
[InputSystem] Fixed typo in CHANGELOG

AlexTyrer committed:
[InputSystem] Fix string check in MatchSinglePropertyContains()

o Was looking for manufacturer string in pattern not pattern in manufacturer string (doh!)

AlexTyrer committed:
[InputSystem] Fix formatting of modified files (used unity-meta format tool)
